### PR TITLE
Add ARC auto-sync architecture

### PR DESCRIPTION
**Pull request by R33v0LT**
**Your checklist for this pull request**
- [x] I've documented or updated the documentation of every API function and struct this PR changes.
- [x] I've added the necessary ARC files for the disassembler to work properly
- [x] I've added ARC to python bindings
- [x] I've added ARC to auto-sync scripts
- [x] Dev fuzzing is done without errors
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)

**Detailed description**

This PR adds support for the ARC architecture in capstone. It is worth noting that this architecture is experimental in lvm, that is, full support for all instructions is not fully implemented. Also, MC tests in llvm (like **llvm_root/llvm/test/MC/Disassembler/ARC/ldst.txt** ) have a format different from other architectures, which is why tests auto-generated by MCUpdater.py does not add instruction bytes to yaml. I think this format will change when ARC is fully supported
To successfully pass the autosync test, a merge of the following PR is required: https://github.com/capstone-engine/llvm-capstone/pull/69

**Test plan**

There are no differences in the way capstone is built with the ARC architecture
MC tests were generated by MCUpdater. Only instruction bytes are manually added. All other tests are added manually
At the moment, all tests are passing successfully, there are CI/CD results to confirm

**Closing issues**

None

### CODE REVIEW

**Comment by R33v0LT** [comment]
@Rot127, fuzzing, as I expected, shows that those instructions that are not yet supported for ARC in llvm lead to incorrect decoding and program crash. llvm-mc from the official github also fails with an error. Should we do something now or ignore it until llvm updates?

**Comment by Rot127** [comment]
You can fix it in our fork of LLVM. And open an issue in LLVM with a reference to the fix commit. They can cherry-pick it. If you have the time, you can also open a PR with the fix yourself.

**Comment by R33v0LT** [comment]
The problem is solved now. The DecodeGPR32RegisterClass return value was not checked (in llvm disassembler also)
I must reran fuzzing again to check if all is good now. If so, I'll marked this PR as ready for review

File: arch/ARC/ARCDisassembler.c [hunk 1]
```diff
@@ -200,7 +200,9 @@ static DecodeStatus DecodeMEMrs9(MCInst *Inst, unsigned Insn, uint64_t Address,
 	// We have the 9-bit immediate in the low bits, 6-bit register in high bits.
 	unsigned S9 = Insn & 0x1ff;
 	unsigned R = (Insn & (0x7fff & ~0x1ff)) >> 9;
-	DecodeGPR32RegisterClass(Inst, R, Address, Decoder);
+	if (!DecodeGPR32RegisterClass(Inst, R, Address, Decoder)) {
```
Comment (Rot127) [inline arch/ARC/ARCDisassembler.c]:
```suggestion
	if (DecodeGPR32RegisterClass(Inst, R, Address, Decoder) == MCDisassembler_Fail) {
```

Please do the check explicitly. We never know if they might decide changing the values.

**Comment by Rot127** [comment]
Do you consider this done? Then I would review it this week.

**Comment by R33v0LT** [comment]
> Do you consider this done? Then I would review it this week.

Yes. I think it is done now

File: arch/ARC/ARCDisassembler.c [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+		}
+		if (BytesLen >= 8) {
+			// Attempt to decode 64-bit instruction.
+			uint64_t Insn64;
+			if (!readInstruction64(Bytes, BytesLen, Address, Size, &Insn64))
+				return MCDisassembler_Fail;
+			Result = decodeInstruction(DecoderTable64, Instr,
+						   Insn64, Address, NULL);
+			if (MCDisassembler_Success == Result) {
+				;
```
Comment (Rot127) [inline arch/ARC/ARCDisassembler.c:430]: Could you an issue please that we should also covert the `LLVM_DEBUG` stuff?
It seems to become more and more common to use.
No need to fix it here, but it nice to track.

File: arch/ARC/ARCGenCSFeatureName.inc
Comment (Rot127) [inline arch/ARC/ARCGenCSFeatureName.inc:1]: If they are empty and unused please remove them.
Also the others.

File: arch/ARC/ARCMapping.c [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+void ARC_printer(MCInst *MI, SStream *O,
+		       void * /* MCRegisterInfo* */ info)
+{
+	MCRegisterInfo *MRI = (MCRegisterInfo *)info;
+	MI->MRI = MRI;
+
+	ARC_LLVM_printInst(MI, MI->address, "", O);
+	// ARC_add_cs_groups(MI);
```
Comment (Rot127) [inline arch/ARC/ARCMapping.c]:
```suggestion
```

File: arch/ARC/ARCMapping.c [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+void ARC_setup_op(cs_arc_op *op)
+{
+	memset(op, 0, sizeof(cs_arc_op));
+	op->type = ARC_OP_INVALID;
+}
+
+void ARC_init_cs_detail(MCInst *MI)
+{
+	if (detail_is_set(MI)) {
```
Comment (Rot127) [inline arch/ARC/ARCMapping.c]: Invert condition and return early please.

File: arch/ARC/ARCMapping.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+// cs_detail related functions
+void ARC_init_cs_detail(MCInst *MI);
+void ARC_set_detail_op_imm(MCInst *MI, unsigned OpNum,
+				 arc_op_type ImmType, int64_t Imm);
+void ARC_add_cs_detail(MCInst *MI, int /* arc_op_group */ op_group,
+			     va_list args);
+static inline void add_cs_detail(MCInst *MI,
+				 int /* arc_op_group */ op_group, ...)
+{
+	if (!MI->flat_insn->detail)
```
Comment (Rot127) [inline arch/ARC/ARCMapping.h]:
```suggestion
	if (!details_is_set(MI))
```

File: include/capstone/arc.h [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+// 	int64_t disp;	    ///< displacement/offset value
+// } arc_op_mem;
+
+/// Instruction operand
+typedef struct cs_arc_op {
+	arc_op_type type;	      //< operand type
+	union {
+		unsigned int reg;     /// register value for REG operand
+		int64_t imm;	      /// immediate value for IMM operand
+		// arc_op_mem mem; /// base/disp value for MEM operand
```
Comment (Rot127) [inline include/capstone/arc.h]: Do you want to add the memory operand in the future? If so we should do it before the v6 Gold release. Because changing the API afterwards is meh.

File: cstool/cstool.c [hunk 7]
```diff
@@ -359,6 +363,11 @@ static void usage(char *prog)
 		printf(")\n");
 	}
 
+	if (cs_support(CS_ARCH_ARC)) {
+		printf("        arc         arc little endian\n");
+		printf("        arcbe       arc big endian\n");
+	}
+
```
Comment (Rot127) [inline cstool/cstool.c]:
```suggestion
```

Isn't necessary anymore, because you added it above. Also, there was no big endian support in `ARCDisassembler.c`

File: bindings/python/capstone/arc_const.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+ARC_INS_STH_DI = 189
+ARC_INS_STH_S = 190
+ARC_INS_STH = 191
+ARC_INS_ST_AB = 192
+ARC_INS_ST_AW = 193
+ARC_INS_ST_DI_AB = 194
+ARC_INS_ST_DI_AW = 195
+ARC_INS_ST_DI = 196
+ARC_INS_ST = 197
+ARC_INS_SUB1_ = 198
```
Comment (Rot127) [inline bindings/python/capstone/arc_const.py]:
```suggestion
ARC_INS_SUB1_ = 198
```

These instruction ids ending in `_` look fishy. I suspect they end in a `.` or with a `{`. But they for sure end with a non alphabetical character. Can you check if this is expected? For some archs a `.` actually belongs to the mnemonic (PPC). For some it doesn't (ARM).

If the mnemonic should not end in a special character, you can check the `getNormalizedMnemonic` function in the `PrinterCapstone`. You can remove special chars there.

File: bindings/python/capstone/arc.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        ('reg', ctypes.c_uint),
+        ('imm', ctypes.c_int64)
+    )
+
+
+class ARCOp(ctypes.Structure):
+    _fields_ = (
+        ('type', ctypes.c_uint),
+        ('value', ARCOpValue),
+        ('access', ctypes.c_uint8)
```
Comment (Rot127) [inline bindings/python/capstone/arc.py]:
```suggestion
        ('access', ctypes.c_int)
```

It was an enum in `arc.h`. Please either make it an `uint8` or change it here to an `int`

File: bindings/python/capstone/arc.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+class ARCOpValue(ctypes.Union):
+    _fields_ = (
+        ('reg', ctypes.c_uint),
+        ('imm', ctypes.c_int64)
+    )
+
+
+class ARCOp(ctypes.Structure):
+    _fields_ = (
+        ('type', ctypes.c_uint),
```
Comment (Rot127) [inline bindings/python/capstone/arc.py]:
```suggestion
        ('type', ctypes.c_int),
```
enums are ints if I am not mistaken.

File: bindings/python/capstone/arc.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    def reg(self):
+        return self.value.reg
+
+
+# Instruction structure
+class CsARC(ctypes.Structure):
+    _fields_ = (
+        ('op_count', ctypes.c_uint8),
+        ('operands', ARCOp * 8),
+        ('update_flags', ctypes.c_bool),
```
Comment (Rot127) [inline bindings/python/capstone/arc.py]:
```suggestion
```

**Review by Rot127 — CHANGES_REQUESTED** [review]
Continue later (at `arc.h`). But so far looks very good!

File: suite/auto-sync/src/autosync/Targets.py [hunk 12]
```diff
@@ -36,8 +38,10 @@
     "LoongArch": "LoongArch",
     "SystemZ": "SystemZ",
     "Mips": "Mips",
+    "ARC": "ARC",
```
Comment (Rot127) [inline suite/auto-sync/src/autosync/Targets.py]:
```suggestion
```

File: suite/auto-sync/src/autosync/cpptranslator/CppTranslator.py [hunk 13]
```diff
@@ -429,10 +433,15 @@ def translate(self) -> None:
...
                 # Here we bundle these captures together.
+                # if "template_declaration" in pattern:
+                #     print(self.tree.root_node)
                 query: Query = self.ts_cpp_lang.query(pattern)
                 captures_bundle: [[(Node, str)]] = list()
                 for q in query.captures(self.tree.root_node):
                     if q[1] == patch.get_main_capture_name():
+                        # from autosync.cpptranslator.patches.Helper import get_text
+                        # if patch.get_main_capture_name() == "template_def":
+                        #     print(get_text(self.src, q[0].start_byte, q[0].end_byte).decode())
```
Comment (Rot127) [inline suite/auto-sync/src/autosync/cpptranslator/CppTranslator.py]:
```suggestion
```

File: suite/auto-sync/src/autosync/cpptranslator/CppTranslator.py [hunk 14]
```diff
@@ -429,10 +433,15 @@ def translate(self) -> None:
                 # Each patch has a capture which includes the whole subtree searched for.
                 # Additionally, it can include captures within this subtree.
                 # Here we bundle these captures together.
+                # if "template_declaration" in pattern:
+                #     print(self.tree.root_node)
```
Comment (Rot127) [inline suite/auto-sync/src/autosync/cpptranslator/CppTranslator.py]:
```suggestion
```

File: suite/auto-sync/src/autosync/cpptranslator/patches/BadConditionCode.py
Comment (Rot127) [inline suite/auto-sync/src/autosync/cpptranslator/patches/BadConditionCode.py:1]: Please add a test for this new Patch in `test_patches.py`.

File: suite/auto-sync/src/autosync/cpptranslator/patches/FieldFromInstr.py [hunk 15]
```diff
@@ -47,14 +47,20 @@ def get_patch(self, captures: [(Node, str)], src: bytes, **kwargs) -> bytes:
         else:
             # Get the Val/Inst parameter.
             # Its type determines the instruction width.
-            inst_param: Node = param_list_caller.named_children[1]
+            if len(param_list_caller.named_children) == 1:
+                # If function just return fieldFromInstruction(...)
+                inst_param: Node = param_list_caller.named_children[0]
+            else:
+                inst_param = param_list_caller.named_children[1]
```
Comment (Rot127) [inline suite/auto-sync/src/autosync/cpptranslator/patches/FieldFromInstr.py:54]: Also, please add a test for this edge case.

File: suite/cstest/include/test_detail_arc.h [hunk 16]
```diff
@@ -0,0 +1,60 @@
+// Copyright © 2024 Rot127 <unisono@quyllur.org>
```
Comment (Rot127) [inline suite/cstest/include/test_detail_arc.h]: Your copyright

File: suite/cstest/src/test_detail_arc.c [hunk 17]
```diff
@@ -0,0 +1,99 @@
+// Copyright © 2024 Rot127 <unisono@quyllur.org>
```
Comment (Rot127) [inline suite/cstest/src/test_detail_arc.c]: Also your copyright please

File: suite/test_corpus3.py [hunk 18]
```diff
@@ -134,6 +135,8 @@ def test_file(fname):
         ("CS_ARCH_ALPHA", "CS_MODE_BIG_ENDIAN"): 56,
         ("CS_ARCH_HPPA", "CS_MODE_HPPA_11+CS_MODE_BIG_ENDIAN"): 57,
         ("CS_ARCH_HPPA", "CS_MODE_HPPA_20+CS_MODE_BIG_ENDIAN"): 58,
+        ("CS_ARCH_ARC", "CS_MODE_LITTLE_ENDIAN"): 59,
+        ("CS_ARCH_ARC", "CS_MODE_BIG_ENDIAN"): 60,
```
Comment (Rot127) [inline suite/test_corpus3.py]:
```suggestion
```

File: tests/details/arc.yaml
Comment (Rot127) [inline tests/details/arc.yaml:1]: Do these tests cover all the possible `OP_GROUPS`? If no, please add one test for each op group.
Otherwise, all good.

Also, please add tests which checks the implicit reads, writes and groups.

**Review by Rot127 — CHANGES_REQUESTED** [review]
This was a very straight forward review! Excellent job, well done!

Please add `ARC` in `cs_v6_release_guide.md` und `New features`.
Simple two liner is enough. Just please mention about the memory operands.

Otherwise, please address the other small requests.

Regarding the not yet implemented memory operand type: Do you want to add them? It complicates the details a bit. But on the other hand there seems only the `MemOperandRI` type. So not too much work I think? Or was there another problem?

Nonetheless, we can merge it like this. But we should have them before v6 Gold.
cc @XVilka @kabeor

File: bindings/python/capstone/arc_const.py (hunk 8 above)
Comment (R33v0LT) [reply bindings/python/capstone/arc_const.py]: It works like that because of instructions like `sub1.$cc`. I can replace last dot if the mnemonic cointains it

File: include/capstone/arc.h (hunk 6 above)
Comment (R33v0LT) [reply include/capstone/arc.h]: There's a problem here: not all instructions that actually use MemRI go through `printMemOperandRI`. As a result, it turns out that, for example, the `ldb %r0, [%r1,4]` instruction will contain operand.mem, while `ldh.ab %r0, [%r1,4]` will not, despite they are similar

File: arch/ARC/ARCDisassembler.c (hunk 2 above)
Comment (R33v0LT) [reply arch/ARC/ARCDisassembler.c:430]: LLVM_DEBUG is already present in the patches

File: bindings/python/capstone/arc_const.py (hunk 8 above)
Comment (Rot127) [reply bindings/python/capstone/arc_const.py]: Yes, please remove it.

File: include/capstone/arc.h (hunk 6 above)
Comment (Rot127) [reply include/capstone/arc.h]: I see. Please open an issue about it. So we can track it.

File: arch/ARC/ARCMapping.c [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+
+	unsigned OpNum = va_arg(args, unsigned);
+	cs_op_type op_type = map_get_op_type(MI, OpNum);
+	cs_op_type base_op_type = op_type;
+	cs_op_type offset_op_type;
+	// Fill cs_detail
+	switch (op_group) {
+	default:
+		printf("ERROR: Operand group %d not handled!\n", op_group);
+		assert(0);
```
Comment (Rot127) [inline arch/ARC/ARCMapping.c]:
```suggestion
		CS_ASSERT_RET(0);
```

File: include/capstone/arc.h [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+	ARC_INS_STH_FAR,
+	ARC_INS_ST_FAR,
+	ARC_INS_ADC_,
+	ARC_INS_ADC_F,
+	ARC_INS_ADC,
+	ARC_INS_ADD_S,
+	ARC_INS_ADD_,
+	ARC_INS_ADD_F,
+	ARC_INS_ADD,
+	ARC_INS_AND_,
```
Comment (Rot127) [inline include/capstone/arc.h]: Please don't forget about those ones.

File: suite/auto-sync/src/autosync/cpptranslator/patches/BadConditionCode.py [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+
+from tree_sitter import Node
+
+from autosync.cpptranslator.patches.Patch import Patch
+
+
+class BadConditionCode(Patch):
+    """
+    Patch   return BadConditionCode
+    to      assert(0 && "Unknown condition code passed")
```
Comment (Rot127) [inline suite/auto-sync/src/autosync/cpptranslator/patches/BadConditionCode.py]:
```suggestion
    to      CS_ASSERT(0 && "Unknown condition code passed")
```

File: suite/auto-sync/src/autosync/cpptranslator/patches/BadConditionCode.py [hunk 22]
```diff
@@ -0,0 +1,32 @@
+# Copyright © 2024 Dmitry Sibitsev <unisono@quyllur.org>
```
Comment (Rot127) [inline suite/auto-sync/src/autosync/cpptranslator/patches/BadConditionCode.py]: Your email please  :)

**Review by Rot127 — CHANGES_REQUESTED** [review]
Nice! Almost done. Just those few nitpicks.

Also, please mark the suggestions as resolved when done.

File: include/capstone/arc.h (hunk 20 above)
Comment (R33v0LT) [reply include/capstone/arc.h]: Created pull request for this to capstone-llvm: https://github.com/capstone-engine/llvm-capstone/pull/72

File: include/capstone/arc.h (hunk 6 above)
Comment (R33v0LT) [reply include/capstone/arc.h]: Do you mean open an issue in capstone?

**Review by Rot127 — APPROVED** [review]

**Comment by notxvilka** [comment]
@R33v0LT there are some conflicts now, could you please rebase it?

**Review by kabeor — APPROVED** [review]
Thank you! It's amazing!

**Comment by Rot127** [comment]
Rebased with https://github.com/capstone-engine/capstone/pull/2614

**Comment by Rot127** [comment]
Closed with https://github.com/capstone-engine/capstone/pull/2614
